### PR TITLE
feat: add clone and remotes to GitRepo

### PR DESCRIPTION
**Pull request by dariuszd21**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

It adds additional methods to the `GitRepo` for repository cloning and managing remotes configuration (creating, renaming and checking for the existence)

(CRAFT-3087)

### CODE REVIEW

File: craft_application/git/_git_repo.py [hunk 1]
```diff
@@ -308,3 +405,46 @@ def _resolve_ref(self, ref: str) -> str:
...
+            raise GitError("Cannot clone to existing repository")
+
+        logger.debug("Cloning %s to %s", url, path)
+        if checkout_branch is not None:
+            logger.debug("Checking out to branch: %s", checkout_branch)
+
+        try:
+            # TODO: consider using git CLI with --single-branch
+            #  as it is not implemented in pygit2
+            #  https://github.com/libgit2/pygit2/issues/1290
```
Comment (lengau) [inline craft_application/git/_git_repo.py:438]: I'm not sure this is worth switching to the CLI version of git for, but I like having the reference to the issue so we can implement that once pygit2 implements it.

Could you make a feature request for `--single-branch`, tag it as blocked, and reference the upstream issue?

File: craft_application/git/_git_repo.py [hunk 2]
```diff
@@ -193,6 +231,65 @@ def _init_repo(self) -> None:
...
+            self._repo.remotes.rename(remote_name, new_remote_name)
+        except KeyError as ke:
+            raise GitError(f"remote `{remote_name}` does not exist.") from ke
+        except ValueError as ve:
+            raise GitError(
+                f"wrong name `{new_remote_name}` for the remote provided."
+            ) from ve
+        except pygit2.GitError as error:
+            raise GitError(
+                f"cannot rename `{remote_name}` to `{new_remote_name}`"
```
Comment (lengau) [inline craft_application/git/_git_repo.py:290]:
```suggestion
                f"cannot rename '{remote_name}' to '{new_remote_name}'"
```

File: craft_application/git/_git_repo.py [hunk 3]
```diff
@@ -193,6 +231,65 @@ def _init_repo(self) -> None:
...
+        :raises GitError: if remote cannot be renamed
+        """
+        logger.debug("Renaming `%s` to `%s`", remote_name, new_remote_name)
+        try:
+            self._repo.remotes.rename(remote_name, new_remote_name)
+        except KeyError as ke:
+            raise GitError(f"remote `{remote_name}` does not exist.") from ke
+        except ValueError as ve:
+            raise GitError(
+                f"wrong name `{new_remote_name}` for the remote provided."
```
Comment (lengau) [inline craft_application/git/_git_repo.py:286]:
```suggestion
                f"wrong name '{new_remote_name}' for the remote provided."
```

File: craft_application/git/_git_repo.py [hunk 4]
```diff
@@ -193,6 +231,65 @@ def _init_repo(self) -> None:
...
+        new_remote_name: str,
+    ) -> None:
+        """Rename remote in the repository configuration.
+
+        :param remote_name: the remote repository name
+        :param new_remote_name: the new name for the remote
+
+        :raises GitError: if remote cannot be renamed
+        """
+        logger.debug("Renaming `%s` to `%s`", remote_name, new_remote_name)
```
Comment (lengau) [inline craft_application/git/_git_repo.py:279]:
```suggestion
        logger.debug("Renaming '%s' to '%s'", remote_name, new_remote_name)
```

File: craft_application/git/_git_repo.py [hunk 5]
```diff
@@ -193,6 +231,65 @@ def _init_repo(self) -> None:
...
+        try:
+            self._repo.remotes.rename(remote_name, new_remote_name)
+        except KeyError as ke:
+            raise GitError(f"remote `{remote_name}` does not exist.") from ke
```
Comment (lengau) [inline craft_application/git/_git_repo.py:283]:
```suggestion
            raise GitError(f"remote '{remote_name}' does not exist.") from ke
```

File: craft_application/git/_git_repo.py [hunk 6]
```diff
@@ -308,3 +405,46 @@ def _resolve_ref(self, ref: str) -> str:
...
+                path=os.fspath(path),
+                checkout_branch=checkout_branch,
+                depth=depth,
+                callbacks=GitLogCallbacks(),
+            )
+        except KeyError as ke:
+            raise GitError(f"cannot find branch `{checkout_branch}` in {url}") from ke
+        except pygit2.GitError as ge:
+            raise GitError(f"cannot clone repository: {url} to {str(path)!r}") from ge
+        return cls(path)
```
Comment (dariuszd21) [inline craft_application/git/_git_repo.py:450]: Rewrite to subprocess, due to performance implications

File: craft_application/git/_git_repo.py [hunk 7]
```diff
@@ -193,6 +231,65 @@ def _init_repo(self) -> None:
...
+
+        :param remote_name: the remote repository name
+        :param new_remote_name: the new name for the remote
+
+        :raises GitError: if remote cannot be renamed
+        """
+        logger.debug("Renaming `%s` to `%s`", remote_name, new_remote_name)
+        try:
+            self._repo.remotes.rename(remote_name, new_remote_name)
+        except KeyError as ke:
```
Comment (lengau) [inline craft_application/git/_git_repo.py:282]: Please use normal convention (either `exc` or `error`) for these names.

**Review by lengau — CHANGES_REQUESTED** [review]
I like this, thanks! Some semantic requests :-)

Also, I think this can be considered a `feat` rather than `chore`

File: craft_application/git/_git_repo.py (hunk 1 above)
Comment (lengau) [reply craft_application/git/_git_repo.py:438]: Resolving due to https://github.com/canonical/craft-application/pull/396#discussion_r1690252171

File: craft_application/git/_git_repo.py (hunk 7 above)
Comment (dariuszd21) [reply craft_application/git/_git_repo.py]: Changed in https://github.com/canonical/craft-application/pull/396/commits/d2f2f7796da819576a9468bc2fa8a6b4612da59a

File: craft_application/git/_git_repo.py [hunk 8]
```diff
@@ -193,6 +231,65 @@ def _init_repo(self) -> None:
                 f"Could not initialize a git repository in {str(self.path)!r}."
             ) from error
 
+    def remote_exist(self, remote_name: str) -> bool:
```
Comment (lengau) [inline craft_application/git/_git_repo.py:234]:
```suggestion
    def remote_exists(self, remote_name: str) -> bool:
```

File: craft_application/git/_git_repo.py [hunk 9]
```diff
@@ -308,3 +405,50 @@ def _resolve_ref(self, ref: str) -> str:
...
+
+    @classmethod
+    def clone_repository(
+        cls,
+        *,
+        url: str,
+        path: Path,
+        checkout_branch: str | None = None,
+        depth: int = 0,
+    ) -> GitRepo:
```
Comment (lengau) [inline craft_application/git/_git_repo.py:417]:
```suggestion
    ) -> Self:
```

**Review by lengau — APPROVED** [review]

File: craft_application/git/_git_repo.py (hunk 9 above)
Comment (dariuszd21) [reply craft_application/git/_git_repo.py]: Done in https://github.com/canonical/craft-application/pull/396/commits/7d210e4d46d63bf131a38f3f583a6aaa31202618

File: craft_application/git/_git_repo.py (hunk 8 above)
Comment (dariuszd21) [reply craft_application/git/_git_repo.py]: Done in https://github.com/canonical/craft-application/pull/396/commits/7d210e4d46d63bf131a38f3f583a6aaa31202618

File: craft_application/git/_git_repo.py (hunk 6 above)
Comment (dariuszd21) [reply craft_application/git/_git_repo.py]: Ended up using `process_run` from craft_parts.utils.os_utils and checking if the function is called with correct arguments in tests

**Review by mr-cal — APPROVED** [review]

**Pull request merged**